### PR TITLE
Validate member attributes

### PR DESCRIPTION
**Pull request by ericstj**
Validate APICompat for attributes on members

### CODE REVIEW

**Comment by ericstj** [comment]
Changes have been merged to Arcade, will remove my manual update.  We will need to merge this to pick up the arcade update.

File: src/System.Data.Common/ref/System.Data.Common.manual.cs [hunk 1]
```diff
@@ -15,6 +15,35 @@ internal class ConstraintConverter { }
...
+        [System.ComponentModel.DefaultValueAttribute(typeof(string))]
+        [System.ComponentModel.RefreshPropertiesAttribute(System.ComponentModel.RefreshProperties.All)]
+        [System.ComponentModel.TypeConverter(typeof(ColumnTypeConverter))]
+        public System.Type DataType { get { throw null; } set { } }
+
+        [System.ComponentModel.TypeConverter(typeof(DefaultValueTypeConverter))]
+        public object DefaultValue { get { throw null; } set { } }
+    }
+    internal class ColumnTypeConverter { }
+    internal class DefaultValueTypeConverter { }
```
Comment (stephentoub) [inline src/System.Data.Common/ref/System.Data.Common.manual.cs]: Having public attributes referring to internal types doesn't cause problems?

File: src/System.Data.OleDb/ref/System.Data.OleDb.cs [hunk 2]
```diff
@@ -244,7 +239,7 @@ public sealed partial class OleDbErrorCollection : System.Collections.ICollectio
     public sealed partial class OleDbException : System.Data.Common.DbException
     {
         internal OleDbException() { }
-        public override int ErrorCode { get { throw null; } }
```
Comment (stephentoub) [inline src/System.Data.OleDb/ref/System.Data.OleDb.cs]: Why was this removed?  Because OleDbException is sealed?

File: src/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs [hunk 3]
```diff
@@ -81,10 +81,10 @@ public partial struct WaitForChangedResult
...
         private object _dummy;
         private int _dummyPrimitive;
-        public System.IO.WatcherChangeTypes ChangeType { get { throw null; } set { } }
-        public string Name { get { throw null; } set { } }
-        public string OldName { get { throw null; } set { } }
-        public bool TimedOut { get { throw null; } set { } }
+        public System.IO.WatcherChangeTypes ChangeType { readonly get { throw null; } set { } }
+        public string Name { readonly get { throw null; } set { } }
+        public string OldName { readonly get { throw null; } set { } }
+        public bool TimedOut { readonly get { throw null; } set { } }
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs]: This is interesting.  The C# compiler is adding `readonly` automatically to auto-prop get accessors.  It's a breaking change to remove `readonly` from a property once it's there, which means if we ever wanted to make the getter mutating, we'd have issues.  I don't see that being a problem in this case, but it's something to think about.

Comment (stephentoub) [reply src/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs]: i.e. moving forward any time we see an auto-prop get accessor replaced by an actual implementation, we need to take a hard look at it.  Presumably at that point our compat tooling will tell us that something that used to be readonly no longer is in the implementation?

Comment (stephentoub) [reply src/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs]: It also technically means that when someone changes a public property's implementation in corefx, e.g.
```C#
private int _foo;
public int Foo { get => _foo; set => _foo = value; }
```
to instead be:
```C#
public int Foo { get; set; }
```
they're actually changing the public surface area, the ref will need to be updated, etc.

cc: @terrajobst

**Review by stephentoub — APPROVED** [review]

File: src/System.Data.Common/ref/System.Data.Common.manual.cs (hunk 1 above)
Comment (ericstj) [reply src/System.Data.Common/ref/System.Data.Common.manual.cs]: Not a problem and something we're already doing, it's necessary from the tooling perspective.  They invoke the internal converters.  If we were to do this now I'd say we should make these public instead.

File: src/System.Data.OleDb/ref/System.Data.OleDb.cs (hunk 2 above)
Comment (ericstj) [reply src/System.Data.OleDb/ref/System.Data.OleDb.cs]: It's moved to the manual file.

File: src/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs (hunk 3 above)
Comment (ericstj) [reply src/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs]: We should catch these changes now.  
I think missing these from the refs is a bug in GenAPI where it's not doing the right thing to generate the readonly keyword all the time.  GenAPI should be able to see the readonly attribute that the compiler added and emit readonly.  I haven't yet root caused that /cc @tannergooding.

Comment (tannergooding) [reply src/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs]: CC. @jaredpar, I believe the compiler team stated its a behavioral change not necessarily a breaking change.

**Review by safern — COMMENTED** [review]
LGTM

File: src/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs (hunk 3 above)
Comment (ericstj) [reply src/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs]: I double checked all the places I had to add `readonly` and can confirm that there is no GenAPI bug, so we should be in a good spot with the added validation.  When it flags things they will be real problems, and running GenAPI can fix them.
